Translate top-level declarations

eklimo opened the pull request:
Translate
- Method declarations, constructor declarations, and anonymous initialization blocks
- Block statements
- Property declarations
- Field declarations

Also
- Implement source locations/line numbers

eklimo committed:
Build method declarations

eklimo committed:
Implement node locations & line numbers

eklimo committed:
Build block statements

eklimo committed:
Build property declarations

zorzella commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt:
```diff
@@ -17,15 +17,30 @@ import com.google.summit.ast.TypeRef
 class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptions) {
+    companion object {
+        /** Name used for [MethodDeclaration]s that are constructors. */
+        private const val JORJE_CONSTRUCTOR_ID = "<init>"
```
I'm not clear as to why you are using "JORJE" in the name of these constants.

eklimo replied:
Many rules expect the names of certain method declarations (namely constructors, static init blocks, and property accessors) to match those produced by Jorje, so I am replacing Summit names with Jorje/PMD names during tree building. I can make this more clear in the documentation. Maybe I'll also replace `JORJE_` with `PMD_` as that represents the purpose better.

aaronhurst-google replied:
I suggest moving the string constant to the PMD ASTMethod, since either its internal methods or its callers also need to know what the magic string is.

eklimo replied:
Currently, rules have these strings hard-coded, but I agree that this would be a better location for them so that rules could be refactored in the future.

zorzella commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt:
```diff
@@ -95,6 +113,12 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
+     * Builds [node] using its default building behavior. Used to build a child of a [Node] in place
+     * of itself.
+     */
+    private fun buildInstead(node: Node?) = build(node, parent = null)
```
it's not clear to my why this method is called "build instead"

eklimo replied:
Yes, this is the main part I wanted feedback on.

During development, I saw a pattern where a node would need to delegate its conversion to one of its children. For example, a `MethodDeclaration` is usually turned into an `ASTMethod` within `buildMethodDeclaration`, but in one case the body of the method (`CompoundStatement` -> `ASTBlockStatement`) should be built in its place.

Originally I expected this to be handled by calling the regular `build` function, e.g.
```kotlin
when {
  node.isAnonymousInitializationCode() && !node.hasKeyword(Keyword.STATIC) -> build(node.body, parent)
  // ...
}
```
but this resulted in an issue where the node would be added twice to its parent, for the following reason:
1. The `ASTBlockStatement` node is returned by `build(node.body, parent)`, is returned as the result of `buildMethodDeclaration`, is assigned to `wrapper` in `build` and is added to `parent`.
2. The `build(node.body, parent)` call that constructs the `ASTBlockStatement` assigns it to its own `wrapper` and also adds it to the same `parent`.

Either of these cases can be disabled to fix the duplication issue, so I set `parent` to `null` in `buildInstead` to disable the latter case. I wanted to provide a name for this operation, so I extracted it into a function. I have no preference for the name as long as it makes the purpose clear.

P.S. I definitely see this design as flawed (that the node is both returned by `build` and `setParent` is called by `build`), but I haven't found an alternative that works better yet. I think this would be a good point for me to continue looking for more alternatives while I haven't added many new nodes.

eklimo replied:
I found another way to factor the building functions that I am more satisfied with. See [`6647c4c`](https://github.com/eklimo/pmd/pull/5/commits/6647c4c59459a34769830c59ad67da095e94b20e). Now `setParent` has to be called intentionally and `parent` doesn't have to be set to `null` confusingly.

eklimo committed:
Build field declarations

eklimo committed:
Refactor building primitives

- Split `setParent` call from main `build` function into new
  `buildAndSetParent`
- Remove list-based `build` function to decrease complexity
- Improve documentation

eklimo committed:
Condense `buildModifiers` calls

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTBlockStatement.java:
```diff
@@ -34,8 +36,10 @@ protected void handleSourceCode(final String source) {
-        // curlyBrace = firstChar == '{';
-        // TODO(b/239648780)
+        SourceLocation loc = node.getSourceLocation();
+        String[] lines = source.split("\n");
```
suggest one or more helper function here. Maybe one useful one would be `SourceLocation.extractFrom(String source)String`

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTBlockStatement.java:
```diff
@@ -4,9 +4,11 @@
-import com.google.summit.ast.Node;
 import net.sourceforge.pmd.annotation.InternalApi;
 
+import com.google.summit.ast.Node;
```
Our style requirement is to have (non-static) imports in a single block (i.e. no whitespace) in ASCII sort order.

If PMD has a style, follow that, otherwise ours.

eklimo replied:
PMD's import order is defined by the [Checkstyle](https://github.com/pmd/build-tools/blob/master/src/main/resources/net/sourceforge/pmd/pmd-checkstyle-config.xml#L89) plugin

aaronhurst-google commented on pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt:
```diff
@@ -56,70 +63,132 @@ class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptio
+     * [newName].
+     */
+    private fun cloneMethod(src: MethodDeclaration, newName: String) =
+        MethodDeclaration(
```
It doesn't seem right that the (PMD) tree builder is creating Summit nodes.

How inconvenient would it be to handle these as special cases inside the pmd.ASTMethod?  Or, if that is a bad solution, what would it look like to push these changes upstream into the original Summit tree?

eklimo replied:
The motivation behind this was to match the `image` of `ASTMethod` to the one expected by rules (which is a hard-coded value based on Jorje). I notice that instead of changing the name directly on the `MethodDeclaration`, we can also modify the `image` getter for the same result:
```java
@Override
public String getImage() {
    if (node.isAnonymousInitializationCode()) {
        return STATIC_INIT_ID;
    } else if (node.isConstructor()) {
        return CONSTRUCTOR_ID;
    } else {
        return node.getId().asCodeString();
    }
}
```
How does this sound?

aaronhurst-google replied:
Sounds great! Thanks for exploring the alternate approach.

eklimo commented:
Rebased onto #6 and #7:
- Implement locations/line numbers in each of [`AbstractApexNode.Single`](https://github.com/eklimo/pmd/pull/5/commits/2e750718a60917ba6eb96d157f19f3e5c9b5857c#diff-a06b7194c7b2ee85c429aa91adb28fc0dacb200fe7ea6c92d2c8d760947d9cddR37) and [`AbstractApexNode.Many`](https://github.com/eklimo/pmd/pull/5/commits/2e750718a60917ba6eb96d157f19f3e5c9b5857c#diff-a06b7194c7b2ee85c429aa91adb28fc0dacb200fe7ea6c92d2c8d760947d9cddR73)
